### ISSUE TITLE
Fix other places with quadratic complexity.

### ISSUE THREAD

**Issue by bangerth** [dealii/dealii#11989]
#11984 made me think whether there are other places with quadratic complexity when looping over cells. A quick search yields the following:
```
> egrep -n -r '!= *.*end\([0-9]' include/ source/
source/distributed/tria.cc:2328:           cell != this->end(0);
source/distributed/tria.cc:2336:               cell != this->end(0);
source/distributed/tria.cc:2492:               cell != this->end(0);
source/distributed/tria.cc:3343:           cell != this->end(0);
source/grid/grid_generator.cc:8018:         cell != volume_mesh.end(0);
source/grid/grid_tools_dof_handlers.cc:1154:    for (; cell_1 != mesh_1.end(0); ++cell_1, ++cell_2)
source/grid/grid_tools_dof_handlers.cc:2139:         cell != mesh.end(0);
source/grid/grid_tools_dof_handlers.cc:2219:         cell != mesh.end(0);
source/grid/grid_tools.cc:1152:         cell != triangulation.end(0);
source/dofs/dof_renumbering.cc:1430:             cell != dof_handler.end(0);
```
The ones in `source/distributed/tria.cc` were fixed in the patch linked to above, but we should probably all replace them in the same way as done in #11984.

**Comment by drwells** [comment]
The `end` iterator doesn't ever change (unless the `Triangulation` changes). Why can't we just cache it somewhere to fix this issue more permanently?

**Comment by bangerth** [comment]
Yes, that is the cheap solution. We have many places where we define `endc` before the loop. It just has to be moved out of the loop condition because there it is evaluated every iteration.

**Comment by drwells** [comment]
I meant that we should make `Triangulation::end()` a constant-time function instead of linear, so we don't have to worry about the loops at all.

**Comment by bangerth** [comment]
The `end()` function is constant time, but `end(level)` is not. We could cache what `end(level)` requires in the `NumberCache`.

**Comment by peterrum** [comment]
> The end() function is constant time, but end(level) is not. We could cache what end(level) requires in the NumberCache.

Out of curiosity: which lines lead to the quadratic complexity? I took a quick look at the code and was not sure what causes the issue. Is it the problem that one has to find the first **used** object on the levels? Caching sounds a good idea!

**Comment by kronbichler** [comment]
While it would be nice to get this fixed eventually, I think this is low on our priority list, so I am moving the milestone for now.
